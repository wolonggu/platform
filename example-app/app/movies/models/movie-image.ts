export interface MovieImage {
  [x: string]: any;
}
