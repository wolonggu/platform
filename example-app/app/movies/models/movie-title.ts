export interface MovieTitle {
  id: number;
  titles: any[];
}
