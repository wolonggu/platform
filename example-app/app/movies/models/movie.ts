export interface Movie {
  [x: string]: any;
}
