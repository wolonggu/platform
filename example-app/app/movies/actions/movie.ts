import { Action } from '@ngrx/store';
import { Movie } from '../models/movie';

export enum MovieActionTypes {
  Search = '[Movie] Search',
  SearchComplete = '[Movie] Search Complete',
  SearchError = '[Movie] Search Error',
  Load = '[Movie] Load',
  Select = '[Movie] Select',
}

/**
 * Every action is comprised of at least a type and an optional
 * payload. Expressing actions as classes enables powerful
 * type checking in reducer functions.
 *
 * See Discriminated Unions: https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
 */
export class Search implements Action {
  readonly type = MovieActionTypes.Search;

  constructor(public payload: string) {}
}

export class SearchComplete implements Action {
  readonly type = MovieActionTypes.SearchComplete;

  constructor(public payload: Movie[]) {}
}

export class SearchError implements Action {
  readonly type = MovieActionTypes.SearchError;

  constructor(public payload: string) {}
}

export class Load implements Action {
  readonly type = MovieActionTypes.Load;

  constructor(public payload: Movie) {}
}

export class Select implements Action {
  readonly type = MovieActionTypes.Select;

  constructor(public payload: string) {}
}

/**
 * Export a type alias of all actions in this action group
 * so that reducers can easily compose action types
 */
export type MovieActions =
  | Search
  | SearchComplete
  | SearchError
  | Load
  | Select;
