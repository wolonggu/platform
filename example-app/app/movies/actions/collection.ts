import { Action } from '@ngrx/store';
import { Movie } from '../models/movie';

export enum CollectionActionTypes {
  AddMovie = '[Collection] Add Movie',
  AddMovieSuccess = '[Collection] Add Movie Success',
  AddMovieFail = '[Collection] Add Movie Fail',
  RemoveMovie = '[Collection] Remove Movie',
  RemoveMovieSuccess = '[Collection] Remove Movie Success',
  RemoveMovieFail = '[Collection] Remove Movie Fail',
  Load = '[Collection] Load',
  LoadSuccess = '[Collection] Load Success',
  LoadFail = '[Collection] Load Fail',

  LoadPopular = '[Collection] LoadPopular',
  LoadPopularSuccess = '[Collection] LoadPopular Success',
  LoadPopularFail = '[Collection] LoadPopular Fail',

  LoadUpcoming = '[Collection] LoadUpcoming',
  LoadUpcomingSuccess = '[Collection] LoadUpcoming Success',
  LoadUpcomingFail = '[Collection] LoadUpcoming Fail',

  LoadTopRated = '[Collection] LoadTopRated',
  LoadTopRatedSuccess = '[Collection] LoadTopRated Success',
  LoadTopRatedFail = '[Collection] LoadTopRated Fail',
}

/**
 * LoadUpcoming Collection Actions
 */
export class LoadTopRated implements Action {
  readonly type = CollectionActionTypes.LoadTopRated;
}

export class LoadTopRatedSuccess implements Action {
  readonly type = CollectionActionTypes.LoadTopRatedSuccess;

  constructor(public payload: Movie[]) {}
}

export class LoadTopRatedFail implements Action {
  readonly type = CollectionActionTypes.LoadTopRatedFail;

  constructor(public payload: any) {}
}

/**
 * LoadUpcoming Collection Actions
 */
export class LoadUpcoming implements Action {
  readonly type = CollectionActionTypes.LoadUpcoming;
}

export class LoadUpcomingSuccess implements Action {
  readonly type = CollectionActionTypes.LoadUpcomingSuccess;

  constructor(public payload: Movie[]) {}
}

export class LoadUpcomingFail implements Action {
  readonly type = CollectionActionTypes.LoadUpcomingFail;

  constructor(public payload: any) {}
}

/**
 * LoadPopular Collection Actions
 */
export class LoadPopular implements Action {
  readonly type = CollectionActionTypes.LoadPopular;
}

export class LoadPopularSuccess implements Action {
  readonly type = CollectionActionTypes.LoadPopularSuccess;

  constructor(public payload: Movie[]) {}
}

export class LoadPopularFail implements Action {
  readonly type = CollectionActionTypes.LoadPopularFail;

  constructor(public payload: any) {}
}

/**
 * Add Movie to Collection Actions
 */
export class AddMovie implements Action {
  readonly type = CollectionActionTypes.AddMovie;

  constructor(public payload: Movie) {}
}

export class AddMovieSuccess implements Action {
  readonly type = CollectionActionTypes.AddMovieSuccess;

  constructor(public payload: Movie) {}
}

export class AddMovieFail implements Action {
  readonly type = CollectionActionTypes.AddMovieFail;

  constructor(public payload: Movie) {}
}

/**
 * Remove Movie from Collection Actions
 */
export class RemoveMovie implements Action {
  readonly type = CollectionActionTypes.RemoveMovie;

  constructor(public payload: Movie) {}
}

export class RemoveMovieSuccess implements Action {
  readonly type = CollectionActionTypes.RemoveMovieSuccess;

  constructor(public payload: Movie) {}
}

export class RemoveMovieFail implements Action {
  readonly type = CollectionActionTypes.RemoveMovieFail;

  constructor(public payload: Movie) {}
}

/**
 * Load Collection Actions
 */
export class Load implements Action {
  readonly type = CollectionActionTypes.Load;
}

export class LoadSuccess implements Action {
  readonly type = CollectionActionTypes.LoadSuccess;

  constructor(public payload: Movie[]) {}
}

export class LoadFail implements Action {
  readonly type = CollectionActionTypes.LoadFail;

  constructor(public payload: any) {}
}

export type CollectionActions =
  | AddMovie
  | AddMovieSuccess
  | AddMovieFail
  | RemoveMovie
  | RemoveMovieSuccess
  | RemoveMovieFail
  | Load
  | LoadSuccess
  | LoadFail
  | LoadPopular
  | LoadPopularSuccess
  | LoadPopularFail
  | LoadUpcoming
  | LoadUpcomingSuccess
  | LoadUpcomingFail
  | LoadTopRated
  | LoadTopRatedSuccess
  | LoadTopRatedFail;
