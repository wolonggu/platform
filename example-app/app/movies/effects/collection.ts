import { Injectable } from '@angular/core';
import { Action } from '@ngrx/store';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Database } from '@ngrx/db';
import { Observable } from 'rxjs/Observable';
import { defer } from 'rxjs/observable/defer';
import { of } from 'rxjs/observable/of';
import { Load } from './../actions/movie';

import {
  CollectionActions,
  LoadFail,
  LoadSuccess,
  LoadPopularFail,
  LoadPopularSuccess,
  LoadUpcomingFail,
  LoadUpcomingSuccess,
  LoadTopRatedFail,
  LoadTopRatedSuccess,
  AddMovieSuccess,
  AddMovieFail,
  CollectionActionTypes,
  RemoveMovie,
  RemoveMovieFail,
  RemoveMovieSuccess,
  AddMovie,
} from './../actions/collection';
import { Movie } from '../models/movie';
import { switchMap, toArray, map, catchError, mergeMap } from 'rxjs/operators';
import { TmdbService } from '../../core/services/tmdb.service';

@Injectable()
export class CollectionEffects {
  /**
   * This effect does not yield any actions back to the store. Set
   * `dispatch` to false to hint to @ngrx/effects that it should
   * ignore any elements of this effect stream.
   *
   * The `defer` observable accepts an observable factory function
   * that is called when the observable is subscribed to.
   * Wrapping the database open call in `defer` makes
   * effect easier to test.
   */
  @Effect({ dispatch: false })
  openDB$: Observable<any> = defer(() => {
    return this.db.open('movies_app');
  });

  @Effect()
  loadPopularCollection$: Observable<Action> = this.actions$.pipe(
    ofType(CollectionActionTypes.LoadPopular),
    switchMap(() =>
      this.tmdbService
        .getMoviesPopular()
        .pipe(
          map((movies: Movie[]) => new LoadPopularSuccess(movies)),
          catchError(error => of(new LoadPopularFail(error)))
        )
    )
  );

  @Effect()
  loadTopRatedCollection$: Observable<Action> = this.actions$.pipe(
    ofType(CollectionActionTypes.LoadTopRated),
    switchMap(() =>
      this.tmdbService
        .getMoviesTopRated()
        .pipe(
          map((movies: Movie[]) => new LoadTopRatedSuccess(movies)),
          catchError(error => of(new LoadTopRatedFail(error)))
        )
    )
  );

  @Effect()
  loadUpcomingCollection$: Observable<Action> = this.actions$.pipe(
    ofType(CollectionActionTypes.LoadUpcoming),
    switchMap(() =>
      this.tmdbService
        .getMoviesUpcoming()
        .pipe(
          map((movies: Movie[]) => new LoadUpcomingSuccess(movies)),
          catchError(error => of(new LoadUpcomingFail(error)))
        )
    )
  );

  @Effect()
  loadCollection$: Observable<Action> = this.actions$.pipe(
    ofType(CollectionActionTypes.Load),
    switchMap(() =>
      this.db
        .query('movies')
        .pipe(
          toArray(),
          map((movies: Movie[]) => new LoadSuccess(movies)),
          catchError(error => of(new LoadFail(error)))
        )
    )
  );

  @Effect()
  addMovieToCollection$: Observable<Action> = this.actions$.pipe(
    ofType(CollectionActionTypes.AddMovie),
    map((action: AddMovie) => action.payload),
    mergeMap(movie =>
      this.db
        .insert('movies', [movie])
        .pipe(
          map(() => new AddMovieSuccess(movie)),
          catchError(() => of(new AddMovieFail(movie)))
        )
    )
  );

  @Effect()
  removeMovieFromCollection$: Observable<Action> = this.actions$.pipe(
    ofType(CollectionActionTypes.RemoveMovie),
    map((action: RemoveMovie) => action.payload),
    mergeMap(movie =>
      this.db
        .executeWrite('movies', 'delete', [movie.id])
        .pipe(
          map(() => new RemoveMovieSuccess(movie)),
          catchError(() => of(new RemoveMovieFail(movie)))
        )
    )
  );

  constructor(
    private actions$: Actions,
    private db: Database,
    private tmdbService: TmdbService
  ) {}
}
