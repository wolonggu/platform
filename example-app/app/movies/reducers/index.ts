import { createSelector, createFeatureSelector } from '@ngrx/store';
import * as fromSearch from './search';
import * as fromMovies from './movies';
import * as fromCollection from './collection';
import * as fromRoot from '../../reducers';

export interface MoviesState {
  search: fromSearch.State;
  movies: fromMovies.State;
  collection: fromCollection.State;
}

export interface State extends fromRoot.State {
  movies: MoviesState;
}

export const reducers = {
  search: fromSearch.reducer,
  movies: fromMovies.reducer,
  collection: fromCollection.reducer,
};

/**
 * A selector function is a map function factory. We pass it parameters and it
 * returns a function that maps from the larger state tree into a smaller
 * piece of state. This selector simply selects the `movies` state.
 *
 * Selectors are used with the `select` operator.
 *
 * ```ts
 * class MyComponent {
 *   constructor(state$: Observable<State>) {
 *     this.moviesState$ = state$.pipe(select(getMoviesState));
 *   }
 * }
 * ```
 */

/**
 * The createFeatureSelector function selects a piece of state from the root of the state object.
 * This is used for selecting feature states that are loaded eagerly or lazily.
 */
export const getMoviesState = createFeatureSelector<MoviesState>('movies');

/**
 * Every reducer module exports selector functions, however child reducers
 * have no knowledge of the overall state tree. To make them usable, we
 * need to make new selectors that wrap them.
 *
 * The createSelector function creates very efficient selectors that are memoized and
 * only recompute when arguments change. The created selectors can also be composed
 * together to select different pieces of state.
 */
export const getMovieEntitiesState = createSelector(
  getMoviesState,
  state => state.movies
);

export const getSelectedMovieId = createSelector(
  getMovieEntitiesState,
  fromMovies.getSelectedId
);

/**
 * Adapters created with @ngrx/entity generate
 * commonly used selector functions including
 * getting all ids in the record set, a dictionary
 * of the records by id, an array of records and
 * the total number of records. This reducers boilerplate
 * in selecting records from the entity state.
 */
export const {
  selectIds: getMovieIds,
  selectEntities: getMovieEntities,
  selectAll: getAllMovies,
  selectTotal: getTotalMovies,
} = fromMovies.adapter.getSelectors(getMovieEntitiesState);

export const getSelectedMovie = createSelector(
  getMovieEntities,
  getSelectedMovieId,
  (entities, selectedId) => {
    return selectedId && entities[selectedId];
  }
);

/**
 * Just like with the movies selectors, we also have to compose the search
 * reducer's and collection reducer's selectors.
 */
export const getSearchState = createSelector(
  getMoviesState,
  (state: MoviesState) => state.search
);

export const getSearchMovieIds = createSelector(
  getSearchState,
  fromSearch.getIds
);
export const getSearchQuery = createSelector(
  getSearchState,
  fromSearch.getQuery
);
export const getSearchLoading = createSelector(
  getSearchState,
  fromSearch.getLoading
);
export const getSearchError = createSelector(
  getSearchState,
  fromSearch.getError
);

/**
 * Some selector functions create joins across parts of state. This selector
 * composes the search result IDs to return an array of movies in the store.
 */
export const getSearchResults = createSelector(
  getMovieEntities,
  getSearchMovieIds,
  (movies, searchIds) => {
    return searchIds.map(id => movies[id]);
  }
);

export const getCollectionState = createSelector(
  getMoviesState,
  (state: MoviesState) => state.collection
);

export const getCollectionLoaded = createSelector(
  getCollectionState,
  fromCollection.getLoaded
);
export const getCollectionLoading = createSelector(
  getCollectionState,
  fromCollection.getLoading
);
export const getCollectionMovieIds = createSelector(
  getCollectionState,
  fromCollection.getIds
);

export const getMovieCollection = createSelector(
  getMovieEntities,
  getCollectionMovieIds,
  (entities, ids) => {
    return ids.map(id => entities[id]);
  }
);

export const isSelectedMovieInCollection = createSelector(
  getCollectionMovieIds,
  getSelectedMovieId,
  (ids, selected) => {
    return ids.indexOf(selected) > -1;
  }
);
