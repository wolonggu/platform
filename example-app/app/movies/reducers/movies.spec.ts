import { reducer } from './movies';
import * as fromMovies from './movies';
import { SearchComplete, Load, Select } from '../actions/movie';
import { Movie, generateMockMovie } from '../models/movie';
import { LoadSuccess } from '../actions/collection';

describe('MoviesReducer', () => {
  const movie1 = generateMockMovie();
  const movie2 = { ...movie1, id: '222' };
  const movie3 = { ...movie1, id: '333' };
  const initialState: fromMovies.State = {
    ids: [movie1.id, movie2.id],
    entities: {
      [movie1.id]: movie1,
      [movie2.id]: movie2,
    },
    selectedMovieId: null,
  };

  describe('undefined action', () => {
    it('should return the default state', () => {
      const result = reducer(undefined, {} as any);

      expect(result).toMatchSnapshot();
    });
  });

  describe('SEARCH_COMPLETE & LOAD_SUCCESS', () => {
    function noExistingMovies(
      action: any,
      moviesInitialState: any,
      initialState: any,
      movies: Movie[]
    ) {
      const createAction = new action(movies);

      const result = reducer(moviesInitialState, createAction);

      expect(result).toMatchSnapshot();
    }

    function existingMovies(action: any, initialState: any, movies: Movie[]) {
      // should not replace existing movies
      const differentMovie2 = { ...movies[0], foo: 'bar' };
      const createAction = new action([movies[1], differentMovie2]);

      const expectedResult = {
        ids: [...initialState.ids, movies[1].id],
        entities: {
          ...initialState.entities,
          [movies[1].id]: movies[1],
        },
        selectedMovieId: null,
      };

      const result = reducer(initialState, createAction);

      expect(result).toMatchSnapshot();
    }

    it('should add all movies in the payload when none exist', () => {
      noExistingMovies(SearchComplete, fromMovies.initialState, initialState, [
        movie1,
        movie2,
      ]);

      noExistingMovies(LoadSuccess, fromMovies.initialState, initialState, [
        movie1,
        movie2,
      ]);
    });

    it('should add only new movies when movies already exist', () => {
      existingMovies(SearchComplete, initialState, [movie2, movie3]);

      existingMovies(LoadSuccess, initialState, [movie2, movie3]);
    });
  });

  describe('LOAD', () => {
    const expectedResult = {
      ids: [movie1.id],
      entities: {
        [movie1.id]: movie1,
      },
      selectedMovieId: null,
    };

    it('should add a single movie, if the movie does not exist', () => {
      const action = new Load(movie1);

      const result = reducer(fromMovies.initialState, action);

      expect(result).toMatchSnapshot();
    });

    it('should return the existing state if the movie exists', () => {
      const action = new Load(movie1);

      const result = reducer(expectedResult, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('SELECT', () => {
    it('should set the selected movie id on the state', () => {
      const action = new Select(movie1.id);

      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Selectors', () => {
    describe('getSelectedId', () => {
      it('should return the selected id', () => {
        const result = fromMovies.getSelectedId({
          ...initialState,
          selectedMovieId: movie1.id,
        });

        expect(result).toMatchSnapshot();
      });
    });
  });
});
