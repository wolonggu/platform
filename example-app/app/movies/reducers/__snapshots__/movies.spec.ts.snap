// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesReducer LOAD should add a single movie, if the movie does not exist 1`] = `
Object {
  "entities": Object {
    "278": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "278",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
  },
  "error": "",
  "ids": Array [
    "278",
  ],
  "loading": false,
  "movies": Array [],
  "selectedMovieId": null,
}
`;

exports[`MoviesReducer LOAD should return the existing state if the movie exists 1`] = `
Object {
  "entities": Object {
    "278": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "278",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
  },
  "ids": Array [
    "278",
  ],
  "selectedMovieId": null,
}
`;

exports[`MoviesReducer SEARCH_COMPLETE & LOAD_SUCCESS should add all movies in the payload when none exist 1`] = `
Object {
  "entities": Object {
    "222": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "222",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
    "278": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "278",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
  },
  "error": "",
  "ids": Array [
    "278",
    "222",
  ],
  "loading": false,
  "movies": Array [],
  "selectedMovieId": null,
}
`;

exports[`MoviesReducer SEARCH_COMPLETE & LOAD_SUCCESS should add all movies in the payload when none exist 2`] = `
Object {
  "entities": Object {
    "222": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "222",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
    "278": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "278",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
  },
  "error": "",
  "ids": Array [
    "278",
    "222",
  ],
  "loading": false,
  "movies": Array [],
  "selectedMovieId": null,
}
`;

exports[`MoviesReducer SEARCH_COMPLETE & LOAD_SUCCESS should add only new movies when movies already exist 1`] = `
Object {
  "entities": Object {
    "222": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "222",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
    "278": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "278",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
    "333": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "333",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
  },
  "ids": Array [
    "278",
    "222",
    "333",
  ],
  "selectedMovieId": null,
}
`;

exports[`MoviesReducer SEARCH_COMPLETE & LOAD_SUCCESS should add only new movies when movies already exist 2`] = `
Object {
  "entities": Object {
    "222": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "222",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
    "278": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "278",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
    "333": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "333",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
  },
  "ids": Array [
    "278",
    "222",
    "333",
  ],
  "selectedMovieId": null,
}
`;

exports[`MoviesReducer SELECT should set the selected movie id on the state 1`] = `
Object {
  "entities": Object {
    "222": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "222",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
    "278": Object {
      "adult": false,
      "backdrop_path": "/xBKGJQsAIeweesB79KC89FpBrVr.jpg",
      "genre_ids": Array [
        "18",
        "80",
      ],
      "id": "278",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "popularity": 6.741296,
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "release_date": "1994-09-10",
      "title": "The Shawshank Redemption",
      "video": false,
      "vote_average": 8.32,
      "vote_count": 5238,
    },
  },
  "ids": Array [
    "278",
    "222",
  ],
  "selectedMovieId": "278",
}
`;

exports[`MoviesReducer Selectors getSelectedId should return the selected id 1`] = `"278"`;

exports[`MoviesReducer undefined action should return the default state 1`] = `
Object {
  "entities": Object {},
  "error": "",
  "ids": Array [],
  "loading": false,
  "movies": Array [],
  "selectedMovieId": null,
}
`;
